### PR TITLE
Fix transient detection for proxy

### LINKED ISSUE #2043: System.Reflection.TargetException when an interface is used as class mapping proxy definition
Hello,

as u may know, we are currently updating from an old NHibernate 4 (4.0.3) version to the current NHibernate 5 (5.2.3) version and are facing some issues.

We often declare an interface as class proxy, as it is stated in the [Mapping Documetation](https://nhibernate.info/doc/nhibernate-reference/mapping.html).

> proxy (optional): Specifies an interface to use for lazy initializing proxies. You may specify the name of the class itself.

With NHibernate 5 a System.Reflection.TargetException is thrown now, when a Many-To-One reference is updated with a proxy object. I first thought, that this only happens in case of union-subclass-mappings (cause we often used proxy definitions in inheritance mappings), but this isn't the case. It is independ from inheritance mappings. Let me explain. 

This is an example model:
```javascript
public class EntityWithInterfaceProxyDefinition: IEntityProxy
{
	public virtual Guid Id { get; set; }
	public virtual string Name { get; set; }
}
````

This is the model where it is referenced:
```javascript
public class EntityWithInterfaceLookup
{
	public virtual Guid Id { get; set; }
	public virtual string Name { get; set; }
	public virtual IEntityProxy EntityLookup { get; set; }
}
```

I've deactivated the reflection optimization to better debug this problem:
```javascript
Cfg.Environment.UseReflectionOptimizer = false;
```

The exception was triggerd from the flush after the update and is thrown here:
```javascript
protected virtual object GetIdentifierPropertyValue(object entity)
{
	return idGetter.Get(entity);
}
```
This makes of course sense, because:
```javascript
entity.GetType().ToString()
"IEntityProxyProxy"
entity.GetType().BaseType.ToString()
"System.Object"
entity.GetType().GetInterfaces()
{System.Type[3]}
    [0]: {Name = "ISerializable" FullName = "System.Runtime.Serialization.ISerializable"}
    [1]: {Name = "INHibernateProxy" FullName = "NHibernate.Proxy.INHibernateProxy"}
    [2]: {Name = "IEntityProxy" FullName = "NHibernate.Test.NHSpecificTest.IlogsProxyTest.IEntityProxy"}
idGetter
{NHibernate.Properties.BasicPropertyAccessor.BasicGetter}
    Method: {System.Guid get_Id()}
    Property: {System.Guid Id}
    PropertyName: "Id"
    ReturnType: {Name = "Guid" FullName = "System.Guid"}
    clazz: {Name = "EntityWithInterfaceProxyDefinition" FullName = "NHibernate.Test.NHSpecificTest.IlogsProxyTest.EntityWithInterfaceProxyDefinition"}
    @property: {System.Guid Id}
    propertyName: "Id"
```
As you can see, the given entity parameter doesn't inherit from the "EntityWithInterfaceProxyDefinition" class, which is the parent class of the idGetter. So you get an TargetException by invoking the idGetter.

I guess this fix raised this issue:
> SHA-1: da56a73706cfed3e186def5538046fbe73bddc19
> * Fix TransientObjectException in ISession.IsDirty() with many-to-one (#1420)
> Add an explicit check for transient objects in ManyToOne.IsDirty
> Fixes #1419

I've attached some test mapping classes, as well as a test fixture with two tests to reproduce this problem. The test "UpdateEntityWithInterfaceLookup" fails the other one doesn't fail (cause it uses the clsas itself as proxy)

I know removing the proxy definition should be the workaround for this issue. But as long you don't say, we are doing something completly wrong, i would rather avoid that.

Thx, in adavance.
[TargetExceptionTest.zip](https://github.com/nhibernate/nhibernate-core/files/2945691/TargetExceptionTest.zip)

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #2043

### CODE REVIEW

File: src/NHibernate/Engine/ForeignKeys.cs [hunk 1]
```diff
@@ -179,9 +179,17 @@ public static bool IsNotTransientSlow(string entityName, object entity, ISession
...
 
 			// let the interceptor inspect the instance to decide
-			// let the persister inspect the instance to decide
-			return session.Interceptor.IsTransient(entity) ??
-			       session.GetEntityPersister(entityName, entity).IsTransient(entity, session);
+
+			if (session.Interceptor.IsTransient(entity) == true)
+				return true;
+
+			if (entity is INHibernateProxy proxy && proxy.HibernateLazyInitializer.IsUninitialized)
```
Comment (bahusoid) [inline src/NHibernate/Engine/ForeignKeys.cs]: Can proxy be at all transient? Can't this check be simplified to `enitity is INHibernateProxy`?

**Comment by bahusoid** [comment]
Oups.. I've accidentally merged it to upstream :(

**Comment by bahusoid** [comment]
Is it OK to hard reset it? Or what should I do?

**Comment by fredericDelaporte** [comment]
I have done it. The branch is protected and cannot be forced pushed to by default. I have also raised the protection.

**Comment by bahusoid** [comment]
@fredericDelaporte Thanks.. I blame Ride for this (just decided to give it a shot) :) And somehow made this push without any confirmations dialog...

**Comment by fredericDelaporte** [comment]
Rider? Never got this issue with it. But my default upstream is never nhibernate repository but mine instead. The additional constraint (review required) I have added for merging PR will not apply if merging locally then pushing with upstream set to nhibernate.

**Comment by bahusoid** [comment]
> Rider? 

Yeah.. This treacherous setting bite me: Show PUSH dialog only when committing to protected branches 
IMHO It shouldn't be ON by default...

**Pull request merged**